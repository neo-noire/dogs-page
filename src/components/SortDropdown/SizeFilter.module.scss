.sizeFilter {
  position: relative;
  width: 200px;
  .popup {
    width: 100%;
    position: absolute;

    display: grid;
    grid-template-rows: 0fr;
    overflow: hidden;
    background-color: #242424;
    border-radius: 0.5rem;
    box-shadow: 1px 1px 6px 0px rgba(255, 255, 255, 0.5);
    transition: grid-template-rows 0.5s ease-in-out;
    ul {
      overflow: hidden;
      padding: 0;
      margin: 0;
      li {
        padding: 0.3rem;
        border-radius: 0.3rem;
        cursor: pointer;
        list-style: none;

        &:hover {
          background-color: rgb(84, 81, 81);
        }
      }
    }

    &.open {
      grid-template-rows: 1fr;
    }
  }
}
