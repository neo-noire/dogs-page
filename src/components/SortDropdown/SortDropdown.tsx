import React, { useState, FC } from "react";
import styles from "./SizeFilter.module.scss";

const sizes: number[] = [10, 25, 50, 75, 100];
interface ISizeFilter<T> {
  controlFunction: (value: any) => void;
  dropdownItem: number | string;
  figure: string;
  sortArray: Array<T>;
}
export const SortDropdown: FC<ISizeFilter<T>> = ({
  controlFunction,
  dropdownItem,
  figure,
  sortArray,
}) => {
  const [open, setOpen] = useState(false);
  return (
    <div className={styles.sizeFilter}>
      <button onClick={() => setOpen(!open)}>
        {figure}: {dropdownItem}
      </button>
      <div className={open ? `${styles.popup} ${styles.open}` : styles.popup}>
        <ul>
          {sortArray.map((el, idx) => (
            <li
              onClick={() => {
                controlFunction(el);
                setOpen(false);
              }}
              key={idx}
              className={
                dropdownItem === (el.name ? el.name : el) ? styles.choosen : ""
              }
            >
              {el.name ? el.name : el}
            </li>
          ))}
        </ul>
      </div>
    </div>
  );
};
