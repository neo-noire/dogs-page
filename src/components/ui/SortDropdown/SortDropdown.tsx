import { useState } from "react";
import styles from "./SizeFilter.module.scss";
import { ISortParams } from "../../../types/types";

export interface ISizeFilter<T> {
  controlFunction: (value: T) => void;
  choosenItem: number | string;
  figure: string;
  sortArray: T[];
}
export const SortDropdown = <T extends ISortParams | number>({
  controlFunction,
  choosenItem,
  figure,
  sortArray,
}: ISizeFilter<T>) => {
  const [open, setOpen] = useState(false);
  return (
    <div className={styles.sizeFilter}>
      <button
        aria-label="sort-open"
        data-testid="sort-open"
        onClick={() => setOpen(!open)}
      >
        {figure}: {choosenItem}
      </button>
      <div
        data-testid="dropdown-list"
        className={open ? `${styles.popup} ${styles.open}` : styles.popup}
      >
        <ul>
          {sortArray.map((el, idx) => (
            <li
              onClick={() => {
                controlFunction(el);
                setOpen(false);
              }}
              key={idx}
              className={
                choosenItem === (typeof el === "number" ? el : el?.name)
                  ? styles.choosen
                  : ""
              }
            >
              {typeof el === "number" ? el : el?.name}
            </li>
          ))}
        </ul>
      </div>
    </div>
  );
};
