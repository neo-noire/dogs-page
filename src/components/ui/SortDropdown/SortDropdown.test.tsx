import { beforeEach, describe, expect, it, vi } from "vitest";
import { fireEvent, render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";
import { SortDropdown } from "./SortDropdown";

describe("Sort Dropdown Component", () => {
  let item = 1;
  const onClick = vi.fn((value: number) => (item = value));
  let button: HTMLButtonElement;
  let dropdownContainer: HTMLDivElement;
  let ul: HTMLUListElement;
  let component: any;
  beforeEach(() => {
    component = render(
      <SortDropdown
        figure="Test"
        sortArray={[1, 2, 3, 4]}
        choosenItem={item}
        controlFunction={onClick}
      />
    );

    button = screen.getByRole("button");
    dropdownContainer = screen.getByTestId(/dropdown-list/i);
    ul = screen.getByRole("list");
  });
  it("It is Rendered", () => {
    expect(screen.getByText(/test/i)).toBeInTheDocument();
    // expect(component).toMatchSnapshot();
  });
  it("It renders correct number of elements", () => {
    expect(ul.childNodes.length).toBe(4);
  });
  it("It renders more then 3 elements", () => {
    expect(ul.childNodes.length).toBeGreaterThan(3);
  });
  it("It renders less then 5 elements", () => {
    expect(ul.childNodes.length).toBeLessThan(5);
  });

  it("class open added on button click", () => {
    expect(dropdownContainer.classList).toMatch(/popup/i);
    fireEvent.click(button);
    expect(dropdownContainer.className).toMatch(/open/i);
  });
  it("class open toggles on button click", () => {
    expect(dropdownContainer.classList).toMatch(/popup/i);
    fireEvent.click(button);
    expect(dropdownContainer.className).toMatch(/open/i);
    fireEvent.click(button);
    expect(dropdownContainer.className).not.toMatch(/open/i);
  });

  it("button text rendered correctly", () => {
    expect(button.textContent).toBe("Test: 1");
  });
  it("li element change button content on click", () => {
    fireEvent.click(button);
    expect(dropdownContainer.className).toMatch(/open/i);
    fireEvent.click(ul.childNodes[1]);
    expect(dropdownContainer.className).not.toMatch(/open/i);

    component.rerender(
      <SortDropdown
        figure="Test"
        sortArray={[1, 2, 3, 4]}
        choosenItem={item}
        controlFunction={onClick}
      />
    );
    expect(button.textContent).toMatch("Test: 2");
  });
});
