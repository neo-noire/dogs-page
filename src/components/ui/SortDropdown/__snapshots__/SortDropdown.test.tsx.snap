// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Sort Dropdown Component > It is Rendered 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="_sizeFilter_c78f7c"
      >
        <button
          aria-label="sort-open"
        >
          Test
          : 
          1
        </button>
        <div
          class="_popup_c78f7c"
          data-testid="dropdown-list"
        >
          <ul>
            <li
              class="_choosen_c78f7c"
            >
              1
            </li>
            <li
              class=""
            >
              2
            </li>
            <li
              class=""
            >
              3
            </li>
            <li
              class=""
            >
              4
            </li>
          </ul>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="_sizeFilter_c78f7c"
    >
      <button
        aria-label="sort-open"
      >
        Test
        : 
        1
      </button>
      <div
        class="_popup_c78f7c"
        data-testid="dropdown-list"
      >
        <ul>
          <li
            class="_choosen_c78f7c"
          >
            1
          </li>
          <li
            class=""
          >
            2
          </li>
          <li
            class=""
          >
            3
          </li>
          <li
            class=""
          >
            4
          </li>
        </ul>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
